src/compiler/checker.ts

```diff
@@ -17528,6 +17574,8 @@ namespace ts {
             const secondAssignableToFirst = isTypeAssignableTo(type2, type1);

```

Comment:
aside: getBestChoiceType is a bad name. "Choice type" doesn't reflect that this is only used in `?:` and `||`. "Conditional result type" is better but would make for a clunky function name.